File: elementary_data_structures/python/linked_list.py
class Element:
    __slots__ = ["data", "next", "prev"]

    def __init__(self, data):
        self.data = data
        self.next = self.prev = None


class LinkedList:
    __slots__ = ["size", "head", "tail"]

    def __init__(self):
        self.size = 0
        self.head = self.tail = None

    def add(self, data):
        self.size += 1

        z = Element(data)

        if not self.head:
            self.head = z
        else:
            self.tail.next = z
        z.prev = self.tail
        self.tail = z

    def remove(self, data):
        z = self.__contains(data)
        if not z:
            print(f"{data} is not found in the list")

            return

        self.size -= 1

        if z.next:
            z.next.prev = z.prev
        else:
            self.tail = z.prev
        if z.prev:
            z.prev.next = z.next
        else:
            self.head = z.next

    def __contains(self, data):
        p = self.head
        while p and p.data != data:
            p = p.next

        return p

    def print_list(self):
        print(f"size = {self.size}", end="")

        if not self.head:
            print(" (empty)")

            return

        print("\n--- from head to tail ---")
        p = self.head
        while p:
            print(p.data, end=" -> " if p.next else "\n")

            p = p.next

        print("--- from tail to head ---")
        p = self.tail
        while p:
            print(p.data, end=" <- " if p.prev else "\n")

            p = p.prev


if __name__ == "__main__":
    ll = LinkedList()

    while True:
        print("1:add 2:remove 3:print > ", end="")
        if (op := int(input())) == 1:
            print("input data > ", end="")
            data = int(input())

            ll.add(data)
        elif op == 2:
            print("input data > ", end="")
            data = int(input())

            ll.remove(data)
        elif op == 3:
            ll.print_list()
        else:
            print("invalid operation")

File: elementary_data_structures/golang/linked_list.go
package main

import "fmt"

type Element struct {
	Data       int
	Next, Prev *Element
}

type LinkedList struct {
	Size       int
	Head, Tail *Element
}

func (ll *LinkedList) add(data int) {
	ll.Size++

	z := &Element{data, nil, nil}

	if ll.Head == nil {
		ll.Head = z
	} else {
		ll.Tail.Next = z
	}
	z.Prev = ll.Tail
	ll.Tail = z
}

func (ll *LinkedList) remove(data int) {
	z := ll.contains(data)

	if z == nil {
		fmt.Printf("%d is not found in the list\n", data)

		return
	}

	ll.Size--

	if z.Next != nil {
		z.Next.Prev = z.Prev
	} else {
		ll.Tail = z.Prev
	}
	if z.Prev != nil {
		z.Prev.Next = z.Next
	} else {
		ll.Head = z.Next
	}
}

func (ll *LinkedList) contains(data int) *Element {
	for p := ll.Head; p != nil; p = p.Next {
		if p.Data == data {
			return p
		}
	}

	return nil
}

func (ll *LinkedList) printList() {
	fmt.Printf("size = %d\n", ll.Size)

	if ll.Head == nil {
		fmt.Println("(empty)")

		return
	}

	fmt.Println("--- from head to tail ---")
	for p := ll.Head; p != nil; p = p.Next {
		fmt.Print(p.Data)
		if p.Next != nil {
			fmt.Print(" -> ")
		} else {
			fmt.Print("\n")
		}
	}

	fmt.Println("--- from tail to head ---")
	for p := ll.Tail; p != nil; p = p.Prev {
		fmt.Print(p.Data)
		if p.Prev != nil {
			fmt.Print(" <- ")
		} else {
			fmt.Print("\n")
		}
	}
}

func main() {
	LL := &LinkedList{}

	for {
		fmt.Print("1:add 2:remove 3:print > ")
		var op int
		fmt.Scanf("%d", &op)

		switch op {
		case 1:
			fmt.Print("input data > ")
			var data int
			fmt.Scanf("%d", &data)
			LL.add(data)
		case 2:
			fmt.Print("input data > ")
			var data int
			fmt.Scanf("%d", &data)
			LL.remove(data)
		case 3:
			LL.printList()
		default:
			fmt.Println("invalid operation")
		}
	}
}
